### ISSUE TITLE
Feature request: Option to recalculate values when changing units

### ISSUE THREAD

**Issue by UlrichSi** [Open-Systems-Pharmacology/PK-Sim#787]
Inspired by #774 : During the beta test phase of PK-Sim 5/MoBi 3, we once were recalculating the values when changing units. Thus, "1 h" would become "60 min" by just switching the unit. This was found to be annoying by most users, but there are situations where this might be desired. 

Could be have an "alternative mode" doing just this by holding down a key like "Shift" or "Alt" while changing the unit via a dropbox?

**Comment by PavelBal** [comment]
Yes, yes, yes, would be great (with an option to switch on/off).

**Comment by StephanSchaller** [comment]
Yes, I would also very much like this behaviour (maybe a global option in the settings?)
For me, it is way more intuitive if I change 1 h to min, that it automatically switches to 60.

In any case this is the behaviour I would have needed every time I changed units. I never had the case, that I just wanted to change units but keep the number.

So maybe this is a personal preference that should be available to be set in the global (user-related, not project related) options.

**Comment by Yuri05** [comment]
Both use cases are justified so it can be done only via a program settings ("_Recalculate value vs. Change **display** unit_"). Agree with @StephanSchaller : this should be a user setting, not a project setting.
Default should be "_Recalculate value_"

**Comment by UlrichSi** [comment]
I am not sure if we want to have "Recalculate" as default. When starting a new project and entering new parameter values, a user typically first enters the number, then changes the unit - what you don't want. To my knowledge it was exactly that situation caused us to disable this function. In particular, if it happens at a training class with new users - so I would recommend to leave the default as it is.

**Comment by StephanSchaller** [comment]
> a user typically first enters the number, then changes the unit

I believe this is user-dependent behaviour ;-) (I personally always select the unit and then enter the value), but I think most people do it the way Ulrich described, so we can leave the default as is.

**Comment by Yuri05** [comment]
>When starting a new project and entering new parameter values, a user typically first enters the number, then changes the unit 

That's exactly what I mean with "_Recalculate value_". If your default unit is e.g. `min `and you would like to enter `2 h`; then you enter the value `2` first (and parameter value becomes `2 min`) and after that you change the unit from `min `to `h` (and parameter value becomes `2h` (or `120 min` - so it's **recalculated** from` 2 min` to `120 min`).
So in short: "_Recalculate value_" = < _Current situation_ > (we might call it differently if not clear)

**Comment by msevestre** [comment]
What about using the Project/user display units it you'd rather see min instead of sec?

**Comment by UlrichSi** [comment]
@Yuri05 did you try that out yourself? Changing a time unit back and forth (h -> min -> h) will currently leave the value in your example on 2 and never recalculates it to 120. Or do we already have an option for that, hidden somewhere deep inside the intestinal tract? ;-)

Anyway - leaving the behavior by default as it is will make most sense.

@msevestre  Using the display unit is not so great since it switches everything. There are cases where you need some times in sec, some in days.

**Comment by StephanSchaller** [comment]
> (we might call it differently if not clear)

Hm, yes, I think I would refer to this as "Change value".

For me "recalculate" means keeping the value but changing display units (changing 2 h to 120 min)

**Comment by UlrichSi** [comment]
For me "recalculate" means keeping the value but changing display units (changing 2 h to 120 min)

**Comment by msevestre** [comment]
I am not sure how often you need to do this kind of unit conversion (because this is just unit conversion that you are describing). If not that often, just use Google;-) 
You can type 56 ml/min in until/h and it does it for you. Do we need to implement it in the tool in that case?

**Comment by UlrichSi** [comment]
Google does not need to know everything we want to convert. ;-)
Also, just for the kicks, try to convert a second order kinetic constant from L/µmol/min to L/mol/sec.
And it also won't work that way if you are offline, like on a plane or a busy train.

Unit conversion is certainly not the most frequently used feature, but is it really that difficult to implement? Since this kind of conversion is done when switching the display units, my simple-minded impression is that everything needed should be already in place, and mostly GUI changes are required.

**Comment by PavelBal** [comment]
> Do we need to implement it in the tool in that case?

Yes!

> You can type 56 ml/min in umol/h

And now convert kcal to the standard energy-unit (I keep forgetting it because its so weird) ;)

**Comment by UlrichSi** [comment]
> And now convert kcal to the standard energy-unit (I keep forgetting it because its so weird) ;)

Strange - we don't even have the energy SI unit J or kJ in our unit list, only our internally used kg-dm-min derivative. Seems everybody sticks to strange legacy units, as being the case for mm Hg pressure.

**Comment by msevestre** [comment]
@PavelBal ok. Too hard for google :)

@UlrichSi Can you add a new feature request somewhere (MoBi or PKSim) to add thoe new SI Units?

**Comment by msevestre** [comment]
>and mostly GUI changes are required.

Those are the hardest changes. We cannot automate tests on those so any UI changes is always difficult.

That being said, I don't think there is any UI changes involved at that stage (aside from the option to be defined in the user settings). The **behavior** of changing the unit value is abstracting deep down and we could change what it does depending on user settings.

That being said, this needs to be done in PKSim and MoBi.
